src/platform/android/java/chip/platform/NsdManagerServiceResolver.java

```diff
@@ -105,62 +108,180 @@ public void run() {
           }
         };
 
-    if (nsdManagerResolverAvailState != null) {
-      nsdManagerResolverAvailState.acquireResolver();
+    NsdServiceFinderAndResolver serviceFinderResolver =
+        new NsdServiceFinderAndResolver(
+            this.nsdManager,
+            serviceInfo,
+            callbackHandle,
+            contextHandle,
+            chipMdnsCallback,
+            timeoutRunnable,
+            multicastLock,
+            mainThreadHandler,
+            nsdManagerResolverAvailState);
+    serviceFinderResolver.start();
+
+    Executors.newSingleThreadScheduledExecutor()
+        .schedule(
+            new Runnable() {
+              @Override
+              public void run() {
+                nsdManager.stopServiceDiscovery(serviceFinderResolver);
```

Comment:
Can you add a log here indicating we're bailing due to a timeout?
```suggestion
Log.e(TAG, "Service discovery timed out after " + BROWSE_SERVICE_TIMEOUT_SEC + " seconds");
nsdManager.stopServiceDiscovery(serviceFinderResolver);
```

Reply:
Done!